### PR TITLE
Fix NPE from MetadataIndexUpgrader

### PR DESCRIPTION
**Pull request by jeeminso**
Noticed an NPE from [wacklig](https://wacklig.pipifein.dev/github/crate/crate/c6de9d54-b441-40f3-b0f6-1168c9ee3ce5/d29c17ec-86f4-41c7-affe-d62d46f2752f/stdout):

```
[2023-06-10T00:07:28,723][WARN ][o.e.c.s.ClusterApplierService] [[cratedb[node_sm1][clusterApplierService#updateTask][T#1]]] failed to notify ClusterStateListener
java.lang.NullPointerException: Cannot invoke "Object.hashCode()" because "pk" is null
	at java.util.ImmutableCollections$MapN.probe(ImmutableCollections.java:1321) ~[?:?]
	at java.util.ImmutableCollections$MapN.get(ImmutableCollections.java:1235) ~[?:?]
	at io.crate.types.DataTypes.ofMappingName(DataTypes.java:514) ~[crate-server.jar:?]
	at io.crate.metadata.upgrade.MetadataIndexUpgrader.addIndexColumnSources(MetadataIndexUpgrader.java:157) ~[crate-server.jar:?]
	at io.crate.metadata.upgrade.MetadataIndexUpgrader.addIndexColumnSources(MetadataIndexUpgrader.java:158) ~[crate-server.jar:?]
	at io.crate.metadata.upgrade.IndexTemplateUpgrader.archiveUnknownOrInvalidSettings(IndexTemplateUpgrader.java:105) ~[crate-server.jar:?]
	at io.crate.metadata.upgrade.IndexTemplateUpgrader.apply(IndexTemplateUpgrader.java:65) ~[crate-server.jar:?]
	at io.crate.metadata.upgrade.IndexTemplateUpgrader.apply(IndexTemplateUpgrader.java:54) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.metadata.TemplateUpgradeService.lambda$new$0(TemplateUpgradeService.java:89) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.metadata.TemplateUpgradeService.calculateTemplateChanges(TemplateUpgradeService.java:220) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.metadata.TemplateUpgradeService.clusterChanged(TemplateUpgradeService.java:125) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:518) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:508) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:476) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:416) ~[crate-server.jar:?]
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:163) ~[crate-server.jar:?]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:259) ~[crate-server.jar:?]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:222) ~[crate-server.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1623) ~[?:?]
```

### CODE REVIEW

File: server/src/main/java/io/crate/metadata/upgrade/MetadataIndexUpgrader.java [hunk 1]
```diff
@@ -154,7 +154,7 @@ public static boolean addIndexColumnSources(Map<String, Object> rootMapping,
             if (ArrayType.NAME.equals(type)) {
                 Map<String, Object> innerMapping = Maps.get(columnProperties, "inner");
                 String innerType = Maps.get(innerMapping, "type");
-                if (ObjectType.UNTYPED.equals(DataTypes.ofMappingName(innerType))) {
+                if (innerType != null || ObjectType.UNTYPED.equals(DataTypes.ofMappingName(innerType))) {
                     updated |= addIndexColumnSources(rootMapping, innerMapping, columnFQN);
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/metadata/upgrade/MetadataIndexUpgrader.java:158]: > I think this (array of objects) should be aligned with object type handling
ie. innerType == null.

Thanks, I put `!=` by mistake.

> Maybe we need a unit test for array of objects with sub-columns to cover NPE case? maybe we can add anotehr column to MappingConstants which provides 5.3 mapping and add another column which is array(object)

It seems like an `array(object)` does not produce `innerType == null`. Also, I am guessing `MetadataIndexUpgrader` gets executed due to some kind of abnormal circumstances, I will look further.

**Comment by jeeminso** [comment]
UPDATE: the NPE was from `test_copy_deep_nested_object_to_partitioned_table_results_in_dynamic_mapping_updates` was confused with a different test failure I was chasing.

File: server/src/main/java/io/crate/metadata/upgrade/MetadataIndexUpgrader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/metadata/upgrade/MetadataIndexUpgrader.java:158]: Looks like `test_copy_deep_nested_object_to_partitioned_table_results_in_dynamic_mapping_updates` covers the test scenario and all we need is to add `innerType == null`, WDYT?

[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/metadata/upgrade/MetadataIndexUpgrader.java]: The test that caught this was `test_copy_deep_nested_object_to_partitioned_table_results_in_dynamic_mapping_updates` so I extracted its mapping and added a kind of sanity test for the upgrader to verify it did not have trouble consuming the deep nested mapping. I thought I should not affect unrelated test, `test_copy_to_migrated_to_sources`.

**Review by BaurzhanSakhariev — APPROVED** [review]
Thx for catching!
